### DISCUSSION TITLE
Retire `master` branch

### DISCUSSION THREAD

**Discussion by joschi** [dropwizard/dropwizard#4796] [Development]
Now that we have multiple release branches which are maintained/developed on in parallel, I don't think it makes sense to keep the `master` branch as a default branch around.

We regularly merged the `master` branch into `release/2.1.x` to keep track of the Dropwizard 2.1.0 release status.

In the past the `master` branch was the _next_ Dropwizard release which is now not true anymore (see also https://github.com/dropwizard/dropwizard/discussions/4720).

The status quo is as follows:

- `release/2.0.x`: Current stable version.
- `release/2.1.x`: Next stable version.
- `release/3.0.x`: Next major version, still on Java EE (`javax.*`) but on Java 11 and Jetty 10.
- `release/4.0.x`: Future release with Jakarta EE (`jakarta.*`), Java 11, Jetty 11.

I suggest making `release/2.1.x` the current default branch and keep `master` around for historic reasons and not to break any external incoming links but add a big note at the top of the README file.

@dropwizard/committers Any opinions?

**Comment by jplock** [comment]
I’m in favor of changing the default branch. What is preventing people from upgrading from 2.0.x to 2.1.x so we have one less branch to maintain? Juggling merges between these branches seems like it’s going to become time consuming unless we figure out a way to get everyone up to 4.0.x as soon as safely possible. 

Is there any major blocker for why a customer won’t be able to upgrade from 2.0.x or 2.1.x to 3.0.x? Maybe we shift efforts to getting 3.0 out, deprecate 2.x and then release a 4.x beta if all of the dependencies can be updated.

**Reply by joschi** [reply]
> What is preventing people from upgrading from 2.0.x to 2.1.x so we have one less branch to maintain?

Dropwizard 2.1.0 hasn't been released yet (except for a number of betas), but I don't see any hard blockers to make it GA soon.

Regarding the upgrade path from 2.0.x to 2.1.x I currently don't remember any complicated things, but we'd need to clean up the release and upgrade notes for that version before its release.

The upgrade 2.1.x to 3.0.x will be more involved but also shouldn't be complicated for most users, see also the discussion in https://github.com/dropwizard/dropwizard/discussions/4720.

**Reply by zUniQueX** [reply]
Getting users to 4.0.x will probably take some time, at least, if we support a major version 3.0.x with Jetty 10. The release of 3.0.x seems to be more complicated than the release of 4.0.x at the moment. So we will have to wait until logback finally will let us upgrade to 3.0.x (at least, if we don't introduce the odd workaround I found in https://github.com/zUniQueX/dropwizard/commit/b5e713b27f7b490ab61c7e78f022896d55321cba). When users have the ability to stay on `javax` dependencies, I fear most of them will do that instead of upgrading their applications.

As we should release 4.0.x in accordance to 3.0.x and not earlier, we will have to maintain all branches at the moment, although the releases of 3.0.x and 4.0.x are currently blocked and it is unclear when we will be able to bring them out.

When we would want to use @jplock's path, we nevertheless _could_ bring 3.0.x soon (but treated as a temporary solution in favor of 4.0.x IMO)

**Reply by pstackle** [reply]
@joschi Since we haven't released 2.1.0 yet, we could just roll those changes into 3.0 and skip 2.1.x entirely, in which case, I think default branch would become `release/3.0.x`. That would match what the Jetty project is doing, they have their default branch configured as `jetty-10.0.x`.  The main things I can think of that are targeted for 2.1.x are the integration of `dropwizard-health` and the Hibernate Validator expression language changes. 

@zUniQueX I'm not sure I follow what you are saying here.
> The release of 3.0.x seems to be more complicated than the release of 4.0.x at the moment.

Wouldn't 4.0.x be blocked by the same Logback changes and actually even more, because I think the PR we are waiting on for Jetty 10
support in Logback doesn't actually handle the Jetty 11 changes required for the `jakarta` namespace?

In general, I agree that version 3.0 (i.e. Jetty 10/`javax`) should be communicated as a temporary/short-term option to ease the migration to version 4.0.

**Reply by zUniQueX** [reply]
> Wouldn't 4.0.x be blocked by the same Logback changes and actually even more, because I think the PR we are waiting on for Jetty 10
support in Logback doesn't actually handle the Jetty 11 changes required for the jakarta namespace?

@pstackle This currently is a very weird topic there. Logback 1.2.x - the current stable version - has a dependency on Jetty 8.x. In the 1.3.x branch they moved to the `jakarta` namespace from `alpha10` to `alpha11` IIRC. To do this, they had to adapt the new API for `HttpFields.Mutable` when upgrading to Jetty 11 and the exception Jetty 10 suffers from isn't thrown in this case.

This means that my dependency PR would actually create a working version of Dropwizard 4. But this doesn't mean we wouldn't suffer from any logback issue. As described by the Jetty team, our current implementation on Jetty 9.x could probably be affected by the 100% CPU bug or by wrong status codes, as the current and not the committed data is processed. This issues will - as far as I know - also persist in the current logback 1.3.x alpha on the `jakarta` namespace.

Tl;dr the exception blocking the Jetty 10 upgrade doesn't block us for Jetty 11 and DW 4, but the current issues need to be resolved for all our versions.

**Reply by joschi** [reply]
> The main things I can think of that are targeted for 2.1.x are the integration of `dropwizard-health` and the Hibernate Validator expression language changes.

@pstackle There are a few minor updates in Dropwizard 2.1.x which I'd like to get out of the door just because it has been cooking for a relatively long time, such as the upgrade to Liquibase 4.x, argparse4j 0.9.x, Hibernate Validator 6.2.x, Dropwizard Metrics 4.2.x, and Jackson 2.13.x.

<details>
<summary>Changes in dropwizard-dependencies between Dropwizard 2.0.x and 2.1.x</summary>

```diff
diff --git dropwizard-dependencies/pom.xml dropwizard-dependencies/pom.xml
index 438b8d6ac..4d7992854 100644
--- dropwizard-dependencies/pom.xml
+++ dropwizard-dependencies/pom.xml
@@ -5,7 +5,7 @@
     <parent>
         <groupId>io.dropwizard</groupId>
         <artifactId>dropwizard-bom</artifactId>
-        <version>2.0.29-SNAPSHOT</version>
+        <version>2.1.0-SNAPSHOT</version>
         <relativePath>../dropwizard-bom</relativePath>
     </parent>
 
@@ -19,7 +19,7 @@
     <properties>
         <!-- External dependencies -->
         <activation-api.version>1.2.2</activation-api.version>
-        <argparse4j.version>0.8.1</argparse4j.version>
+        <argparse4j.version>0.9.0</argparse4j.version>
         <byte-buddy.version>1.12.8</byte-buddy.version>
         <caffeine.version>2.9.3</caffeine.version>
         <checker-qual.version>3.21.2</checker-qual.version>
@@ -32,13 +32,14 @@
         <freemarker.version>2.3.31</freemarker.version>
         <guava.version>31.0.1-jre</guava.version>
         <hibernate-core.version>5.6.5.Final</hibernate-core.version>
-        <hibernate-validator.version>6.1.7.Final</hibernate-validator.version>
+        <hibernate-validator.version>6.2.2.Final</hibernate-validator.version>
         <hk2.version>2.6.1</hk2.version>
         <httpclient.version>4.5.13</httpclient.version>
         <httpcore.version>4.4.15</httpcore.version>
-        <jackson.version>2.10.5.20201202</jackson.version>
+        <jackson.version>2.13.1</jackson.version>
         <jakarta.annotation-api.version>1.3.5</jakarta.annotation-api.version>
         <jakarta.el.version>3.0.4</jakarta.el.version>
+        <jakarta.inject-api.version>1.0.5</jakarta.inject-api.version>
         <jakarta.servlet-api.version>4.0.4</jakarta.servlet-api.version>
         <jakarta.validation-api.version>2.0.2</jakarta.validation-api.version>
         <jakarta.ws.rs-api.version>2.1.6</jakarta.ws.rs-api.version>
@@ -47,16 +48,16 @@
         <jaxb-api.version>2.3.3</jaxb-api.version>
         <jcip-annotations.version>1.0</jcip-annotations.version>
         <jdbi3.version>3.27.2</jdbi3.version>
-        <jersey.version>2.33</jersey.version>
+        <jersey.version>2.35</jersey.version>
         <jetty-setuid-java.version>1.0.4</jetty-setuid-java.version>
         <jetty.version>9.4.45.v20220203</jetty.version>
         <joda-time.version>2.10.13</joda-time.version>
         <jsr305.version>3.0.2</jsr305.version>
-        <liquibase-core.version>3.10.3</liquibase-core.version>
-        <liquibase-slf4j.version>2.0.0</liquibase-slf4j.version>
+        <liquibase-core.version>4.8.0</liquibase-core.version>
+        <liquibase-slf4j.version>4.0.0</liquibase-slf4j.version>
         <logback-throttling-appender.version>1.1.8</logback-throttling-appender.version>
         <logback.version>1.2.10</logback.version>
-        <metrics4.version>4.1.30</metrics4.version>
+        <metrics4.version>4.2.8</metrics4.version>
         <mustache-compiler.version>0.9.10</mustache-compiler.version>
         <nullaway.version>0.9.5</nullaway.version>
         <slf4j.version>1.7.36</slf4j.version>
@@ -161,18 +162,21 @@
                 <groupId>jakarta.activation</groupId>
                 <artifactId>jakarta.activation-api</artifactId>
                 <version>${activation-api.version}</version>
-                <scope>runtime</scope>
             </dependency>
             <dependency>
                 <groupId>jakarta.annotation</groupId>
                 <artifactId>jakarta.annotation-api</artifactId>
                 <version>${jakarta.annotation-api.version}</version>
             </dependency>
+            <dependency>
+                <groupId>jakarta.inject</groupId>
+                <artifactId>jakarta.inject-api</artifactId>
+                <version>${jakarta.inject-api.version}</version>
+            </dependency>
             <dependency>
                 <groupId>jakarta.xml.bind</groupId>
                 <artifactId>jakarta.xml.bind-api</artifactId>
                 <version>${jaxb-api.version}</version>
-                <scope>runtime</scope>
             </dependency>
             <dependency>
                 <groupId>jakarta.servlet</groupId>
@@ -383,11 +387,6 @@
                 <artifactId>aopalliance-repackaged</artifactId>
                 <version>${hk2.version}</version>
             </dependency>
-            <dependency>
-                <groupId>org.glassfish.hk2.external</groupId>
-                <artifactId>jakarta.inject</artifactId>
-                <version>${hk2.version}</version>
-            </dependency>
 
             <!-- Jersey -->
             <dependency>
```
</details>

@dropwizard/committers Any changes you _absolutely_ want to see in Dropwizard 2.1.x?

**Reply by zUniQueX** [reply]
I think #3580 shouldn't be hard to integrate, so we could bring it in 2.1.0. I personally would appreciate #4198, but both PRs aren't critical and we could ship 2.1.0 without them.

**Reply by zUniQueX** [reply]
> @dropwizard/committers Any changes you _absolutely_ want to see in Dropwizard 2.1.x?

@joschi I just looked what we currently have in the PRs. For the milestone only #3580 and #4213 are open yet, but those are no blockers for the release. The first PR is a nice-to-have feature and the second one should have a different implementation before shipping this. The other PRs aren't really critical either.

If we want to bring the dependency updates for 2.1.0, we could make this a GA IMO. This would give users the time to migrate away from all deprecated code. As 3.0.x will probably take some time, I'll put a vote for bringing a quick 2.2.x, if 3.0.x remains blocked.
